### PR TITLE
Support nullable pandas dtypes

### LINKED ISSUE #3604: Support nullable pandas dtypes
It would be nice to support pandas nullable dtypes:

```
In [7]: import pandas as pd

In [8]: df = pd.DataFrame({"a": pd.Series([1, 2, None])}, dtype="Int8")

In [9]: df
Out[9]:
      a
0     1
1     2
2  <NA>
```

Right now, hypothesis tries to convert the input dtype in `past.column` to a numpy dtype, which won't work with these new dtypes.

### PR DESCRIPTION
**Pull request by Cheukting**
Support the new Pandas nullable dtypes ([see here](https://pandas.pydata.org/docs/user_guide/integer_na.html)).

Closes #3604

### CODE REVIEW

File: hypothesis-python/RELEASE.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+RELEASE_TYPE: minor
+
+This add support nullable pandas dtypes in :func:`~hypothesis.extra.pandas` (:issue:`3604`).
```
Comment (Zac-HD) [inline hypothesis-python/RELEASE.rst:3]:
```suggestion
This release adds support for `nullable pandas dtypes <https://pandas.pydata.org/docs/user_guide/integer_na.html>`__ 
in :func:`~hypothesis.extra.pandas` (:issue:`3604`).  
Thanks to Cheuk Ting Ho for implementing this at the PyCon sprints!
```

File: hypothesis-python/src/hypothesis/extra/pandas/impl.py [hunk 2]
```diff
@@ -92,13 +104,33 @@ def elements_and_dtype(elements, dtype, source=None):
...
             "here.  See https://stackoverflow.com/q/74355937 for workaround patterns."
         )
-    dtype = try_convert(np.dtype, dtype, "dtype")
+
+    pd_dtype_map = {
+        t.name: t for t in getattr(IntegerDtype, "__subclasses__", lambda: [])()
+    }
+
+    # is_na_dtype = isinstance(dtype, str) and dtype[0].isupper()
+    is_na_dtype = False
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/extra/pandas/impl.py:113]: I'd prefer to assign this in each clause of the if/elif/else statement below, and delete the stale comment.

File: hypothesis-python/src/hypothesis/extra/pandas/impl.py [hunk 3]
```diff
@@ -92,13 +104,33 @@ def elements_and_dtype(elements, dtype, source=None):
...
+
+    if issubclass(type(dtype), IntegerDtype):
+        is_na_dtype = True
+        dtype = np.dtype(dtype.name.lower())
+    elif isinstance(dtype, str) and dtype in pd_dtype_map.keys():
+        is_na_dtype = True
+        dtype = pd_dtype_map[dtype]()
+        dtype = np.dtype(dtype.name.lower())
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/extra/pandas/impl.py:121]: This seems redundant?  I think you either mean to assign one to `np_dtype` (but we'd unconditionally overwrite that later), or we can just use

```suggestion
        dtype = np.dtype(dtype.lower())
```

File: hypothesis-python/tests/pandas/test_series.py [hunk 4]
```diff
@@ -61,3 +63,23 @@ def test_unique_series_are_unique(s):
...
+    e = find_any(st, lambda s: s.isna().any())
+    assert type(e.dtype) == pandas.core.arrays.integer.Int8Dtype
+
+
+def test_pandas_nullable_types_class():
+    with pytest.raises(
+        InvalidArgument, match="Otherwise it would be treated as dtype=object"
+    ):
+        st = pdst.series(dtype=pandas.core.arrays.integer.Int8Dtype)
+        find_any(st, lambda s: s.isna().any())
```
Comment (Zac-HD) [inline hypothesis-python/tests/pandas/test_series.py:79]: This test could instead be a new case in one of the big parameter lists in `test_argument_validation.py` (a sibling file).

File: hypothesis-python/src/hypothesis/extra/pandas/impl.py (hunk 2 above)
Comment (Cheukting) [reply hypothesis-python/src/hypothesis/extra/pandas/impl.py]: Done

File: hypothesis-python/tests/pandas/test_series.py (hunk 4 above)
Comment (Cheukting) [reply hypothesis-python/tests/pandas/test_series.py]: Done

File: hypothesis-python/src/hypothesis/extra/pandas/impl.py [hunk 5]
```diff
@@ -92,13 +104,34 @@ def elements_and_dtype(elements, dtype, source=None):
...
+    }
+
+    dtype = pd_dtype_map.get(dtype, dtype)
+
+    if isinstance(dtype, IntegerDtype):
+        is_na_dtype = True
+        dtype = np.dtype(dtype.name.lower())
+    elif dtype in pd_dtype_map:
+        is_na_dtype = True
+        dtype = np.dtype(dtype.lower())
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/extra/pandas/impl.py:119]: This clause is redundant now 🙂

File: hypothesis-python/tests/pytest/test_profiles.py [hunk 6]
```diff
@@ -9,8 +9,8 @@
 # obtain one at https://mozilla.org/MPL/2.0/.
 
 import pytest
-from _hypothesis_pytestplugin import LOAD_PROFILE_OPTION
 
+from _hypothesis_pytestplugin import LOAD_PROFILE_OPTION
```
Comment (Zac-HD) [inline hypothesis-python/tests/pytest/test_profiles.py:13]: We'll want to skip these `shed` changes.

**Review by Zac-HD — COMMENTED** [review]
We're very close!

**Review by Zac-HD — APPROVED** [review]
🚀

**Comment by Cheukting** [comment]
Thanks, @Zac-HD Any more for me to work on?

**Comment by Zac-HD** [comment]
I'd suggest https://github.com/Zac-HD/shed/issues, particularly 17 would be nice!

**Review by Dezzj — COMMENTED** [review]
Veiwed

**Pull request merged**